File: backend/apps/opportunities/serializers/public/opportunity.py
from rest_framework import serializers
from drf_spectacular.utils import extend_schema_field
from drf_spectacular.types import OpenApiTypes
from apps.opportunities.models import Opportunity, SavedOpportunity


class PublicOpportunitySerializer(serializers.ModelSerializer):
    field = serializers.CharField(
        source="field.name",
        read_only=True,
    )

    organisation_logo = serializers.SerializerMethodField()
    is_saved = serializers.SerializerMethodField()

    class Meta:
        model = Opportunity
        fields = [
            "id",
            "title",
            "slug",
            "description",
            "responsibilities",
            "requirements",
            "skills_required",
            "benefits",
            "opportunity_type",
            "organisation",
            "organisation_logo",
            "application_url",
            "location",
            "field",
            "deadline",
            "is_remote",
            'is_saved',
            "created_at",
        ]

    @extend_schema_field(OpenApiTypes.URI)
    def get_organisation_logo(self, obj):
        recruiter_profile = getattr(obj.posted_by, 'recruiter_profile', None)
        if recruiter_profile:
            return recruiter_profile.logo
        return None

    @extend_schema_field(OpenApiTypes.URI)
    def get_is_saved(self, obj):
        request = self.context.get('request')
        if not request or not request.user.is_authenticated:
            return False

        if request.user.role != 'student':
            return False

        return SavedOpportunity.objects.filter(
            student=request.user,
            opportunity=obj,
        ).exists()

File: frontend/src/types/opportunity.ts
export type OpportunityTag = "New" | "Featured" | "Closing soon";

export interface Opportunity {
  id: string;
  title: string;
  slug: string;
  description: string;
  responsibilities: string[];
  requirements: string[];
  skills_required: string[];
  benefits: string[]
  opportunity_type: string;
  organisation: string;
  organisation_logo: string;
  application_url: string;
  location: string;
  field: string;
  deadline: string;
  is_remote: boolean;
  is_saved: boolean;
  created_at: string;
  tag?: OpportunityTag;
  logo?: string;
  company?: string;
  workMode?: string;
  type?: string;
  commitment?: string;
}

export interface OpportunityQueryParams {
  page?: number;
  page_size?: number;
  search?: string;
  opportunity_type?: string;
  field?: string;
  organisation?: string;
  location?: string;
  is_remote?: boolean;
}

export interface PaginatedOpportunityResponse {
  count: number;
  next: string | null;
  previous: string | null;
  results: Opportunity[];
}

export interface PaginatedOpportunities {
  data: Opportunity[];
  meta: {
    page: number;
    perPage: number;
    total: number;
    totalPages: number;
  };
}